src/v2/Controllers/ControllerGeneric.sol

```diff
@@ -0,0 +1,368 @@
+
+        uint256 premiumTVLAfterFee = premiumTVL - premiumFee;
+        uint256 collateralTVLAfterFee = collateralTVL + premiumTVLAfterFee;
+
+        // strike price is not reached so premium is entiled to 0
+        premiumVault.setClaimTVL(_epochId, 0);
+        // strike price is not reached so collateral is entitled to collateralTVL + premiumTVLAfterFee
+        collateralVault.setClaimTVL(_epochId, collateralTVLAfterFee);
+
+        // send premium fees to treasury and remaining TVL to collateral vault
+        premiumVault.sendTokens(_epochId, premiumFee, treasury);
+        // strike price reached so collateral is entitled to collateralTVLAfterFee
+        premiumVault.sendTokens(
+            _epochId,
+            premiumTVLAfterFee,
+            address(collateralVault)
+        );
+
+        // TODO: timestamp removal?
+        emit EpochResolved(
+            _epochId,
+            _marketId,
+            VaultTVL(collateralTVLAfterFee, collateralTVL, 0, premiumTVL),
+            false,
+            block.timestamp,
+            0
+        );
+    }
+
+    /** @notice Trigger epoch invalid when one vault has 0 TVL
+     * @param _marketId Target market index
+     * @param _epochId End of epoch set for market
+     */
+    function triggerNullEpoch(uint256 _marketId, uint256 _epochId) public {
+        (
+            IVaultV2 premiumVault,
+            IVaultV2 collateralVault
+        ) = _checkNullEpochConditions(_marketId, _epochId);
+
+        //set claim TVL to final TVL if total assets are 0
+        if (premiumVault.totalAssets(_epochId) == 0) {
+            premiumVault.resolveEpoch(_epochId);
+            collateralVault.resolveEpoch(_epochId);
+
+            premiumVault.setClaimTVL(_epochId, 0);
+            collateralVault.setClaimTVL(
+                _epochId,
+                collateralVault.finalTVL(_epochId)
+            );
+
+            collateralVault.setEpochNull(_epochId);
+        } else if (collateralVault.totalAssets(_epochId) == 0) {
+            premiumVault.resolveEpoch(_epochId);
+            collateralVault.resolveEpoch(_epochId);
+
+            premiumVault.setClaimTVL(_epochId, premiumVault.finalTVL(_epochId));
+            collateralVault.setClaimTVL(_epochId, 0);
+
+            premiumVault.setEpochNull(_epochId);
+        } else revert VaultNotZeroTVL();
+
+        // TODO: timestamp removal?
+        emit NullEpoch(
+            _epochId,
+            _marketId,
+            VaultTVL(
+                collateralVault.claimTVL(_epochId),
+                collateralVault.finalTVL(_epochId),
+                premiumVault.claimTVL(_epochId),
+                premiumVault.finalTVL(_epochId)
+            ),
+            block.timestamp
+        );
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                                INTERNAL
+    //////////////////////////////////////////////////////////////*/
+    function _checkLiquidationConditions(
```

Comment:
@0xnyxos looking through the internal functions, it seems that most of them have overlapping check conditions, wouldn't it make make code cleaner if we can have `MarketDoesNotExist`, `EpochNotExist` and `EpochFinishedAlready` in a seperate function or even modifier wdyt?